### ISSUE TITLE
2.10 parse error

### ISSUE THREAD

**Issue by ming-codes** [eslint/eslint#6158]
**What version of ESLint are you using?**

2.10

**What parser (default, Babel-ESLint, etc.) are you using?**

Tried both default and babel-eslint

**Please show your full configuration:**

```
{
	"extends": "ember",

	"parser": "babel-eslint",

	"ecmaFeatures": {
		modules: true,
		blockBindings: true,
		arrowFunctions: true,
		objectLiteralShorthandMethods: true,
		objectLiteralShorthandProperties: true,
		templateStrings: true
	},

	"rules": {
		"quote-props": [2, "consistent-as-needed"],
		"no-ternary": 0,
		"newline-per-chained-call": 0,
		"arrow-body-style": 0,
		"no-warning-comments": 0,
		"max-len": [1, 200],
		"camelcase": 0,
		"space-before-function-paren": [1, "never"],
		"one-var": [ 1, "never" ],
		"indent": [2, 2, {SwitchCase: 1}],
		"object-shorthand": [ 2, "always" ],
		"prefer-const": 0,
		"comma-dangle": 0,
		"spaced-comment": 1,
		"object-curly-spacing": 0,
		"arrow-spacing": [ 1, {
			before: true,
			after: true
		} ],
		"no-restricted-syntax": 0,

		"no-console": 0,
		"no-magic-numbers": 0,
		"valid-jsdoc": [2, {
			"requireReturn": false
		}]
	},

	"globals": {
		"Ember": true,
		"$": true,
		"Edgar": true,
		"visit": true,
		"andThen": true,
		"find": true,
		"findWithAssert": true,
		"click": true,
		"triggerEvent": true,
		"wait": true,
		"ok": true,
		"equal": true,
		"notEqual": true,
		"deepEqual": true,
		"strictEqual": true,
		"throws": true,
		"fillIn": true,
		"exists": true,
		"currentPath": true,
		"currentURL": true,
		"currentRouteName": true,
		"breadcrumbValidator": true,
		"renderProfiler": true,
		"expect": true,
		"_": true,
		"test": true,
		"module": true,
		"YAMP": true,
		"App": true
	}
}

```

**What did you do? Please include the actual source code causing the issue.**

```js
test('Create New Rule', async function() {
});
```

**What did you expect to happen?**

No parse error

**What actually happened? Please include the actual, raw output from ESLint.**

```
/Workspace/tests/acceptance/rule-test.js
  1:31  error  Parsing error: Unexpected token function
```

**Comment by guyellis** [comment]
Just upgraded to 2.10. I'm hitting this same error.

**Comment by 311chaos** [comment]
Same here. Latest build picked up 2.10. I am now getting 2 errors. Both seem to be related to usage of `async`.

```export default async context => {.... }```

and 

```export default async (context) => { ...} ```

**Comment by gyandeeps** [comment]
@guyellis @ming-codes Does same setup works with `1.9.x` of eslint?

I think this is more of an issue with `babel-eslint` as we dont support `async` parsing right out of the box.

**Comment by guyellis** [comment]
@gyandeeps yes. works on 2.9.x. The [previous Travis build](https://travis-ci.org/guyellis/plant/builds/130137262) on the project I linked had no lint errors. The only change (from greenkeeper) was to change the eslint version to 2.10.0. [This is the commit that changed eslint](https://github.com/guyellis/plant/commit/3999a57eb54ba38bb55cf554e4052e5741b7e38a) that kicked off that build.

**Comment by gyandeeps** [comment]
Found the issue: Currently its not loading the external parser rather using the default (espree) parser.
Regression introduced here : https://github.com/eslint/eslint/pull/6108

**Comment by gyandeeps** [comment]
@guyellis @ming-codes Guys can u try the solution on here #6160

**Comment by nzakas** [comment]
The mistake is here: https://github.com/eslint/eslint/pull/6108/files#diff-f7bf2b665273dfa66ffa4ad7c0a52bb9R252

`parser` or `parserOptions` shouldn't be provided in that object literal if they are null or undefined.

**Comment by 311chaos** [comment]
I tested #6160 changes against 2.10.0, and it resolved the issues that I was seeing.

**Comment by gyandeeps** [comment]
Thanks for testing @311chaos

**Comment by SpainTrain** [comment]
Also tested https://github.com/eslint/eslint/pull/6160 and solved the issue for me as well.

**Comment by egoroof** [comment]
I'm the same here. Will it be short release 2.10.1 with the fix or it will come in 2.11?

**Comment by mysticatea** [comment]
Sorry about this.
We have a plan to release v2.10.1 within this weekend.

**Comment by guyellis** [comment]
@gyandeeps apologies for the tardy response. Yes that fixes it. I just [referenced the current tip](https://github.com/guyellis/plant/commit/5d9ce6fa5241c5351fd9710cc0ac2abe2f384aed) of the master branch in ESLint and that fixes the build. Travis is complaining about the direct reference to GitHub however everything is building and passing localy.

You and the others working on this are awesome! Thank you! :smile:

**Comment by bebraw** [comment]
Are you certain this was fixed fully?

I can confirm the cli works correctly now but I'm still seeing https://github.com/MoOx/eslint-loader/issues/92 and Gulp guys seem to have problems too https://github.com/adametry/gulp-eslint/issues/153 .

**Comment by leeyeh** [comment]
@bebraw I'm using grunt-eslint and this issue remains. The CLI works fine now.

**Comment by avindra** [comment]
Just updated to `2.10.1` and still seeing unwarranted:

` 1:1  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'`

In webpack build

**Comment by MoOx** [comment]
https://github.com/eslint/eslint/issues/6182
